planner/core/logical_plan_builder.go

```diff
@@ -197,6 +198,108 @@ func (a *aggOrderByResolver) Leave(inNode ast.Node) (ast.Node, bool) {
+	for _, expr := range distinctGbyExprs {
+		// distinct group expr has been resolved in resolveGby.
+		proj.Exprs = append(proj.Exprs, expr)
+
+		// add the newly appended names.
+		if c, ok := expr.(*expression.Column); ok {
+			names = append(names, buildExpandFieldName(c, names[p.Schema().ColumnIndex(c)], ""))
+		} else {
+			names = append(names, buildExpandFieldName(expr, nil, ""))
+		}
+
+		// since we will change the nullability of source col, proj it with a new col id.
+		col := &expression.Column{
+			UniqueID: b.ctx.GetSessionVars().AllocPlanColumnID(),
+			// clone it rather than using it directly,
+			RetType: expr.GetType().Clone(),
+		}
+
+		projSchema.Append(col)
+		distinctGbyCols = append(distinctGbyCols, col)
+	}
+	proj.SetSchema(projSchema)
+	proj.SetChildren(p)
+	// since expand will ref original col and make some change, do the copy in executor rather than ref the same chunk.column.
+	proj.AvoidColumnEvaluator = true
+	proj.Proj4Expand = true
+	newGbyItems := expression.RestoreGbyExpression(distinctGbyCols, gbyExprsRefPos)
+
+	// build expand.
+	rollupGroupingSets := expression.RollupGroupingSets(newGbyItems)
+	// eg: <a,b,c> with rollup => {},{a},{a,b},{a,b,c}
+	// for every grouping set above, we should individually set those not-needed grouping-set col as null value.
+	// eg: let's say base schema is <a,b,c,d>, d is unrelated col, keep it real in every grouping set projection.
+	// 		for grouping set {a,b,c}, project it as: [a,    b,    c,    d,   gid]
+	//      for grouping set {a,b},   project it as: [a,    b,    null, d,   gid]
+	//      for grouping set {a},     project it as: [a,    null, null, d,   gid]
+	// 		for grouping set {},      project it as: [null, null, null, d,   gid]
+	expandSchema := proj.Schema().Clone()
+	expression.AdjustNullabilityFromGroupingSets(rollupGroupingSets, expandSchema)
+	expand := LogicalExpand{
+		rollupGroupingSets: rollupGroupingSets,
+		distinctGroupByCol: distinctGbyCols,
+		// fill the gen col names when building level projections.
+	}.Init(b.ctx, b.getSelectOffset())
+
+	// if we want to use bitAnd for the quick computation of grouping function, then the maximum capacity of num of grouping is about 64.
+	expand.GroupingMode = tipb.GroupingMode_ModeBitAnd
+	if len(expand.rollupGroupingSets) > 64 {
+		expand.GroupingMode = tipb.GroupingMode_ModeNumericSet
+	}
+
+	expand.distinctSize, expand.rollupGroupingIDs, expand.rollupID2GIDS = expand.rollupGroupingSets.DistinctSize()
+	hasDuplicateGroupingSet := len(expand.rollupGroupingSets) != expand.distinctSize
+	// append the generated column for logical Expand.
+	tp := types.NewFieldType(mysql.TypeLonglong)
+	tp.SetFlag(mysql.UnsignedFlag | mysql.NotNullFlag)
+	gid := &expression.Column{
+		UniqueID: b.ctx.GetSessionVars().AllocPlanColumnID(),
+		RetType:  tp,
+		OrigName: "gid",
+	}
+	expand.GID = gid
+	expandSchema.Append(gid)
+	expand.GeneratedColNames = append(expand.GeneratedColNames, gid.OrigName)
```

Comment:
BTE,  just my curiosity, how to hidden this column in result ?

Reply:
GeneratedColNames is used in tiflash column name, tidb won't use it